server/src/main/java/org/candlepin/guice/I18nProvider.java

```diff
@@ -35,54 +31,52 @@
 /**
  * I18nProvider
  */
+@Singleton
 public class I18nProvider extends CommonI18nProvider implements Provider<I18n> {
     private static Logger log = LoggerFactory.getLogger(I18nProvider.class);
     private static Map<Locale, I18n> cache = new ConcurrentHashMap<>();
-
-    private I18n i18n;
+    private final Provider<HttpServletRequest> request;
 
     @Inject
-    public I18nProvider(Injector injector) {
-        HttpServletRequest request = null;
+    public I18nProvider(Provider<HttpServletRequest> request) {
+        this.request = request;
+    }
+
+    @Override
+    public I18n get() {
         Locale locale = null;
 
         try {
-            request = injector.getInstance(HttpServletRequest.class);
+            locale = request.get().getLocale();
         }
         catch (ProvisionException e) {
             // This can happen in pinsetter, or anything else not in an http
             // request.
             // just ignore it.
         }
 
-        if (request != null) {
-            locale = request.getLocale();
-        }
-
         locale = (locale == null) ? Locale.US : locale;
 
         // If the locale does not exist, xnap is pretty inefficient.
         // This cache will hold the records more efficiently.
-        //
-        // Make sure to keep the access wrapped in synchronized so we can
-        // share across threads!
-        synchronized (cache) {
-            i18n = cache.get(locale);
-            if (i18n == null) {
-                log.debug("Getting i18n engine for locale {}", locale);
 
-                i18n = I18nFactory.getI18n(getClass(), getBaseName(), locale, I18nFactory.FALLBACK);
-                cache.put(locale, i18n);
+        // see https://en.wikipedia.org/wiki/Double-checked_locking
```

Comment:
I read that Wikipedia page and it seems like there are a lot of subtle things that can go awry with double-check locking in [Java](https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java).  Specifically the 

```
1.  Thread A notices that the value is not initialized, so it obtains the lock
    and begins to initialize the value.
2.  Due to the semantics of some programming languages, the code generated by the compiler is 
    allowed to update the shared variable to point to a partially constructed object before A has
    finished performing the initialization. For example, in Java if a call to a constructor has been
    inlined then the shared variable may immediately be updated once the storage has been
    allocated but before the inlined constructor initializes the object.[6]
3.  Thread B notices that the shared variable has been initialized (or so it appears), and returns
    its value. Because thread B believes the value is already initialized, it does not acquire the
    lock. If B uses the object before all of the initialization done by A is seen by B (either because
    A has not finished initializing it or because some of the initialized values in the object have
    not yet percolated to the memory B uses (cache coherence)), the program will likely crash.
```

This scenario may not apply to use since we are storing `i18n` in the `cache` map rather than in a field (thus sidestepping step 2), but I just wanted to make sure it's something we've considered.

See also http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html and the "Best Practices for Lazy Initialization" section of http://www.oracle.com/technetwork/articles/javase/bloch-effective-08-qa-140880.html

Personally, I like this part:

**Q**: What's most important to understand about best practices for lazy initialization?

**A**: The single most important piece of advice is "Don't do it unless you need to." 

Why do we need to lazy initialize this cache at all?  We know what locales we support.  Why not populate the map in a static block and be done with it?

Or we could just not do it ourselves and use https://github.com/google/guava/wiki/CachesExplained with a CacheLoader.

If we don't do any of that, I think we should be using an immutable map from Guava or an unmodifiable map from Collections and either map should be marked volatile instead of the ConcurrentHashMap.  See https://stackoverflow.com/a/23807722/6124862

Reply:
Due to the operations we do inside the synch block, I think we're good on points 2 and 3 -- we don't have an object to insert into the map until the constructor returns, so both cases of partial initialization are avoided here.

However, that aside, the point you made about static init rather than JIT is a valid one that comes down to a question of which locals we support today and what the memory cost of loading all of them in straight away is in the general case (probably a wash in IT, potentially significant in Sat).

Reply:
@Ceiu Our translation bundles are currently ~2MB on disk. Static init for 2 MB does not seem like too much in the grand scheme of things when our overall memory footprint is 1GB+ in production systems.

Reply:
@awood @Ceiu @barnabycourt I agree that we should initialize all locales at once during application initialization, but I'd like to tackle that in a separate PR later on, just to keep this PR scoped to just scope changes (pun intended).